# HG changeset patch
# User Cédric Krier <ced@b2ck.com>
# Date 1790785811 -7200
#      Wed Sep 30 18:30:11 2026 +0200
# Node ID 7d7d0ccfb5ab052e3a83b8d33c581d41dcf9eb38
# Parent  d5ace112f8a9315a6740ef75c2bda4cc8b94b90e
Validate stdnum code before compacting and formatting it

Closes #15017

diff --git a/tryton/modules/party/party.py b/tryton/modules/party/party.py
--- a/tryton/modules/party/party.py
+++ b/tryton/modules/party/party.py
@@ -438,12 +438,15 @@
         else:
             for country in eu_vat.guess_country(text):
                 if 'eu_vat' in identifier_types:
-                    code = eu_vat.compact(country + text)
-                    add('eu_vat', code)
+                    code = country + text
+                    if eu_vat.is_valid(code):
+                        code = eu_vat.compact(code)
+                        add('eu_vat', code)
                 elif f'{country}_vat' in identifier_types:
                     vat = get_cc_module(country, 'vat')
-                    code = vat.compact(text)
-                    add(f'{country}_vat', code)
+                    if vat.is_valid(text):
+                        code = vat.compact(text)
+                        add(f'{country}_vat', code)
         return result
 
     @classmethod
@@ -956,18 +959,20 @@
     @fields.depends('type', 'code')
     def on_change_with_code(self):
         code = self.code
-        if self.type and '_' in self.type:
+        if code and self.type and '_' in self.type:
             if module := get_cc_module(*self.type.split('_', 1)):
-                if hasattr(module, 'compact'):
-                    try:
-                        code = module.compact(code)
-                    except stdnum.exceptions.ValidationError:
-                        pass
-                if hasattr(module, 'format'):
-                    try:
-                        code = module.format(code)
-                    except stdnum.exceptions.ValidationError:
-                        pass
+                if (hasattr(module, 'is_valid')
+                        and module.is_valid(code)):
+                    if hasattr(module, 'compact'):
+                        try:
+                            code = module.compact(code)
+                        except stdnum.exceptions.ValidationError:
+                            pass
+                    if hasattr(module, 'format'):
+                        try:
+                            code = module.format(code)
+                        except stdnum.exceptions.ValidationError:
+                            pass
         return code
 
     def pre_validate(self):
@@ -1000,12 +1005,14 @@
                 if module := get_cc_module(*type.split('_', 1)):
-                    if hasattr(module, 'format'):
-                        try:
-                            values['code'] = module.format(code)
-                        except stdnum.exceptions.ValidationError:
-                            pass
-                    if hasattr(module, 'compact'):
-                        try:
-                            values['code_compact'] = module.compact(code)
-                        except stdnum.exceptions.ValidationError:
-                            pass
+                    if (hasattr(module, 'is_valid')
+                            and module.is_valid(code)):
+                        if hasattr(module, 'format'):
+                            try:
+                                values['code'] = module.format(code)
+                            except stdnum.exceptions.ValidationError:
+                                pass
+                        if hasattr(module, 'compact'):
+                            try:
+                                values['code_compact'] = module.compact(code)
+                            except stdnum.exceptions.ValidationError:
+                                pass
         return values
@@ -1017,12 +1024,14 @@
             code_compact = getattr(self, 'code_compact', None)
             if (type := getattr(self, 'type', None)) and '_' in type and code:
                 if module := get_cc_module(*type.split('_', 1)):
-                    if hasattr(module, 'format'):
-                        code = module.format(code)
-                    if hasattr(module, 'compact'):
-                        code_compact = module.compact(code)
-                    else:
-                        code_compact = code
+                    if (hasattr(module, 'is_valid')
+                            and module.is_valid(code)):
+                        if hasattr(module, 'format'):
+                            code = module.format(code)
+                        if hasattr(module, 'compact'):
+                            code_compact = module.compact(code)
+                        else:
+                            code_compact = code
                     if getattr(self, 'code', None) != code:
                         values['code'] = code
                     if getattr(self, 'code_compact', None) != code_compact:
